storage/remote/storage.go

```diff
@@ -62,37 +63,49 @@ func (s *Storage) ApplyConfig(conf *config.Config) error {
 
 	// Update write queues
 
-	newQueues := []*QueueManager{}
+	newQueues := make([]*QueueManager, 0, len(s.queues))
+	oldQueues := s.queues
 	// TODO: we should only stop & recreate queues which have changes,
 	// as this can be quite disruptive.
 	for i, rwConf := range conf.RemoteWriteConfigs {
-		c, err := NewClient(i, &ClientConfig{
-			URL:              rwConf.URL,
-			Timeout:          rwConf.RemoteTimeout,
-			HTTPClientConfig: rwConf.HTTPClientConfig,
-		})
+		ok := i < len(oldQueues)
+		if ok {
+			qm := oldQueues[i]
+			if reflect.DeepEqual(conf.GlobalConfig.ExternalLabels, qm.externalLabels) &&
+				reflect.DeepEqual(rwConf, qm.rwConf) {
```

Comment:
Ditto; isf golang equality not enough? 

> Struct values are comparable if all their fields are comparable. Two struct values are equal if their corresponding non-blank fields are equal.

https://golang.org/ref/spec#Comparison_operators

Reply:
Golang equality for structs only works if there are no pointers or slices in the structs, which there are here. We could do a hash of all the fields, but then we have to maintain a hash function that is always updated to contain all the latest fields (including sub fields like relabel configs and HTTP client configs). So `reflect.DeepEqual()` is probably the best option. And as this is not a hot path...